controller/snapshot_controller.go

```diff
@@ -0,0 +1,585 @@
+}
+
+func syncSnapshotWithSnapshotInfo(snap *longhorn.Snapshot, snapInfo *longhorn.SnapshotInfo, restoreSize int64) error {
+	size, err := strconv.ParseInt(snapInfo.Size, 10, 64)
+	if err != nil {
+		return err
+	}
+	snap.Status.Parent = snapInfo.Parent
+	snap.Status.Children = snapInfo.Children
+	snap.Status.MarkRemoved = snapInfo.Removed
+	snap.Status.UserCreated = snapInfo.UserCreated
+	snap.Status.CreationTime = snapInfo.Created
+	snap.Status.Size = size
+	snap.Status.Labels = snapInfo.Labels
+	snap.Status.RestoreSize = restoreSize
+	if snap.Status.MarkRemoved || snap.DeletionTimestamp != nil {
+		snap.Status.ReadyToUse = false
+	} else {
+		snap.Status.ReadyToUse = true
+		snap.Status.Error = ""
+	}
+	return nil
+}
+
+func (sc *SnapshotController) getTheOnlyEngineCRforSnapshot(snapshot *longhorn.Snapshot) (*longhorn.Engine, error) {
+	engines, err := sc.ds.ListVolumeEngines(snapshot.Spec.Volume)
+	if err != nil {
+		return nil, errors.Wrap(err, "getTheOnlyEngineCRforSnapshot")
+	}
+	if len(engines) != 1 {
+		return nil, fmt.Errorf("getTheOnlyEngineCRforSnapshot: found more than 1 engines for volume %v", snapshot.Spec.Volume)
+	}
+	var engine *longhorn.Engine
+	for _, e := range engines {
+		engine = e
+		break
+	}
+	return engine, nil
+}
+
+func (sc *SnapshotController) handleSnapshotCreate(snapshot *longhorn.Snapshot, engine *longhorn.Engine) error {
+	client, err := GetClientForEngine(engine, sc.engineClientCollection, engine.Status.CurrentImage)
+	if err != nil {
+		return err
+	}
+	snapshotInfo, err := client.SnapshotGet(snapshot.Name)
+	if err != nil {
+		return err
+	}
+	if snapshotInfo == nil {
+		sc.logger.Infof("creating snapshot %v of volume %v", snapshot.Name, snapshot.Spec.Volume)
+		_, err = client.SnapshotCreate(snapshot.Name, snapshot.Spec.Labels)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// handleSnapshotDeletion reaches out to engine process to check and delete the snapshot
+func (sc *SnapshotController) handleSnapshotDeletion(snapshot *longhorn.Snapshot, engine *longhorn.Engine) error {
+	client, err := GetClientForEngine(engine, sc.engineClientCollection, engine.Status.CurrentImage)
+	if err != nil {
+		return err
+	}
+	snapshotInfo, err := client.SnapshotGet(snapshot.Name)
+	if err != nil {
+		return err
+	}
+	if snapshotInfo == nil {
+		return nil
+	}
+
+	if !snapshotInfo.Removed {
+		sc.logger.Infof("deleting snapshot %v", snapshot.Name)
+		if err = client.SnapshotDelete(snapshot.Name); err != nil {
+			return err
+		}
+	}
+	// TODO: Check if the purge failure is handled somewhere else
+	purgeStatus, err := client.SnapshotPurgeStatus()
+	if err != nil {
+		return errors.Wrap(err, "failed to get snapshot purge status")
+	}
+	isPurging := false
+	for _, status := range purgeStatus {
+		if status.IsPurging {
+			isPurging = true
+			break
+		}
+	}
+	if !isPurging {
+		sc.logger.Infof("start SnapshotPurge to delete snapshot %v", snapshot.Name)
+		if err := client.SnapshotPurge(); err != nil {
```

Comment:
We may need to add a backoff/retry interval here. (As a TODO)
Considering that some snapshots cannot be purged immediately. This will lead to lots of meaningless purge calls to the engine.

Reply:
Could you give more detail about the steps that could lead to lots of meaningless purge calls to the engine?

From testing I see that when we delete the snapshot next to volume-head -> snapshot controller get the snapshot -> check to see if the snapshot as been marked as removed -> do a snapshot purge call -> done progressing the snapshot. Even though the snapshot is still present, we don't repeatedly do snapshot purge as there is no resync for this controller. We only retry calling snapshot purge when there is something change to this snapshot (e.g., it has new children which is no longer volume-head)